Add a states migrator helper method and migrate states for New Zealand and Ukraine

Konamiman opened the pull request:
-   [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

https://github.com/woocommerce/woocommerce/pull/35011 and https://github.com/woocommerce/woocommerce/pull/35493 have converted the state codes for New Zealand and Ukraine to the CLDR standard, but no change happened in the database records that were holding these codes, leading to issues in the following areas:

* Shipping locations
* Orders
* The store address

This pull request adds a `MigrationHelper::migrate_country_states` method that performs all the required database changes, and two database migrations to do the changes for New Zealand and Ukraine.

`MigrationHelper::migrate_country_states` will only migrate a maximum of `100*count(states)` records per table (orders meta table if it exists, and posts meta table), this number can be changed via the `woocommerce_migrate_country_states_for_orders_batch_size` filter.

Closes #35535.

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

### How to test the changes in this Pull Request:

1. Switch to WooCommerce 7.0.
2. Create shipping locations (Settings - Shipping) for states in New Zealand and Ukraine.
3. Set the store country/state (Settings - General) to a state of New Zealand or Ukraine.
4. Create a few orders with shipping and billing addresses in New Zealand and Ukraine.
5. Switch to the branch of this pull request and notice how now:
  * The shipping locations you had created appear as an empty list (just commas).
  * The store country appears as Afghanistan.
  * The orders appear as having no associated state.
6. Run the database update.
7. If you have a big number of orders in the database you'll need to wait for multiple Action Scheduler runs (or force it manually).
8. Verify that shipping locations, store address and orders are now correct (with the updated state names).

Both the orders meta table (if it exists) and the posts meta table are updated, and no change is made to the update date field of the orders nor posts table. Thus the orders sync mechanism is not affected, and which is the authoritative table for orders is irrelevant.

Note that for orders only the authoritative table is changed. Once the database is updated there should be pending syncs for the affected orders (in Settings - Advanced - Custom data stores) and once the sync finishes all the data should be adjusted in all the affected tables (`wp_wc_order_addresses` and `wp_postmeta`, the later for the `_billing_state` and `_shipping_state` keys).

If it's helpful for testing the database update can also be forced with the following cli commands:

```
wp option set woocommerce_db_version 7.0.0
wp eval 'WC_Install::run_manual_database_update();'
wp action-scheduler run
```

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

Konamiman committed:
Add the MigrationHelper::migrate_country_states method

...and use it to migrate the db records for New Zealand and Ukraine
to CLDR codes.

Konamiman committed:
Add changelog file

peterfabian commented on plugins/woocommerce/includes/wc-update-functions.php:
```diff
@@ -18,12 +18,16 @@
 use Automattic\WooCommerce\Internal\ProductDownloads\ApprovedDirectories\Synchronize as Download_Directories_Sync;
+use Automattic\WooCommerce\Internal\Utilities\DatabaseUtil;
+use Automattic\WooCommerce\Utilities\OrderUtil;
+use Automattic\WooCommerce\Utilities\StringUtil;
```
Seems like these 3 are unused here.

Konamiman replied:
Yep, these were relevant at the beginning but then I moved the code to another class.

peterfabian commented on plugins/woocommerce/src/Database/Migrations/MigrationHelper.php:
```diff
@@ -75,4 +77,163 @@ public static function generate_on_duplicate_statement_clause( array $columns ):
+			self::migrate_country_states_for_cot_orders( $country_code, $old_to_new_states_mapping );
+		} else {
+			self::migrate_country_states_for_post_orders( $country_code, $old_to_new_states_mapping );
+		}
```
Just thinking out loud, but what happens when there's a race condition between the check and the update?

If the HPOS was disabled and it got enabled in between the test and code execution, the code updates the CPT data store. If the sync is off, the update will be unsynchronized in the CPT tables until the sync is turned on. 

If the HPOS was enabled and it got disabled in between, the code updates HPOS tables. If the sync is off, the CPT tables will still have incorrect data, as only HPOS tables would get updated.

Makes me wonder if we shouldn't just update both locations (depending on whether the HPOS db tables exist). Or trigger sync manually?

Konamiman replied:
I've since changed the approach completely: now I update both tables and don't touch the update dates.

peterfabian commented on plugins/woocommerce/src/Database/Migrations/MigrationHelper.php:
```diff
@@ -75,4 +77,163 @@ public static function generate_on_duplicate_statement_clause( array $columns ):
+				$current_date_gmt,
+				$address_data['post_id']
+			);
+			$wpdb->query( $update_query );
```
It would be nice to comment on why we're doing this. My assumption is to imitate a CRUD transaction that saves the parent order as well, right?

Konamiman replied:
It was to mark the record as updated so that the sync engine would detect it and update the backup record, but that's no longer relevant.

peterfabian commented on plugins/woocommerce/src/Database/Migrations/MigrationHelper.php:
```diff
@@ -75,4 +77,163 @@ public static function generate_on_duplicate_statement_clause( array $columns ):
+		// phpcs:disable WordPress.DB.PreparedSQL.NotPrepared
+		$addresses_data = $wpdb->get_results( $sql, ARRAY_A );
+
+		foreach ( $addresses_data as $address_data ) {
```
If I understand correctly, this will run one SELECT and `2n` UPDATEs where `n` = the number of affected orders.

I wonder if it would be possible to run `c` SELECT and `c` UPDATES where `c` = the number of states within a country. Each UPDATE would update the meta_value of all rows for the same state (across all orders). I think `c << n` most of the time, so it seems more favorable. 

Also, have you tested how this behaves if the number of affected orders is large (e.g. 1m)? I guess it will time out and fail, right?

Konamiman replied:
Again, no longer relevant after the recent code changes (I'm no longer querying the orders/posts tables, only the meta tables).

peterfabian commented on plugins/woocommerce/src/Database/Migrations/MigrationHelper.php:
```diff
@@ -75,4 +77,163 @@ public static function generate_on_duplicate_statement_clause( array $columns ):
+		);
+		$addresses_data = $wpdb->get_results( $sql, ARRAY_A );
+
+		foreach ( $addresses_data as $address_data ) {
```
Same applies to time complexity as below for posts.

peterfabian commented on plugins/woocommerce/src/Database/Migrations/MigrationHelper.php:
```diff
@@ -75,4 +77,163 @@ public static function generate_on_duplicate_statement_clause( array $columns ):
+			}
+		}
+
+		// phpcs:enable WordPress.DB.PreparedSQL.NotPrepared
```
Can this exclusion apply only to the first SELECT?

Konamiman replied:
Nope, PHPCS was complaining about it in all the queries of the method.

peterfabian reviewed (changes requested):
Thanks, this works well and seems to fix all the problematic instances I found. I have a couple of points we can discuss.

In addition to the ones added in the context of the changed files, it makes me think whether we should include the mapping from some past updates while we're at it, e.g. for [Bolivia](https://github.com/woocommerce/woocommerce/commit/dd64b5c8ff81668d0683adab06e63d077e94885f), [Guatemala](https://github.com/woocommerce/woocommerce/commit/59910ffeab164249bf6cbaea0bbae5b1b6014ff5) or [Venezuela](https://github.com/woocommerce/woocommerce/commit/2bf383a6436600bbe419dc8a811d22fc7b8c8092).

Konamiman committed:
Change the approach for the states migration:

- Don't change order update dates since we are doing a one-time data
  patching that should be transparent to the order syncing engine.
- Account for databases with a high volume of orders by doing the
  updates in batches. The batch size is by default
  200*count(states) when COT tables exist and 100*count(states)
  when not, but this can be customized via the new
  woocommerce_migrate_country_states_for_orders_batch_size filter.
- Register one separate update callback per country.

peterfabian commented on plugins/woocommerce/src/Database/Migrations/MigrationHelper.php:
```diff
@@ -130,110 +136,116 @@ private static function migrate_country_states_for_shipping_locations( string $c
+			"
+			SELECT 1 FROM {$wpdb->prefix}postmeta
+			WHERE (meta_key='_billing_state' OR meta_key='_shipping_state')
+			AND meta_value IN {$states_as_comma_separated}
```
Just curious, is there a reason why this var is interpolated while the `$country_code` is `prepare()`-d? Is it because country code can contain `%`?

Konamiman replied:
The country code is a simple string, it works fine with `prepare` and `%s`. `$states_as_comma_separated`, however, is a comma-separated and parenthesized list of strings, e.g. `('NL', 'WA', 'AK',...)` and needs to be interpolated; it's the same case as `{$wpdb->prefix}`.

peterfabian replied:
Ah yeah of course, it would treat it as one string instead of SQL expression with multiple strings.

peterfabian commented on plugins/woocommerce/src/Database/Migrations/MigrationHelper.php:
```diff
@@ -130,110 +136,116 @@ private static function migrate_country_states_for_shipping_locations( string $c
+				"
+			SELECT EXISTS(
+				SELECT 1 FROM {$wpdb->prefix}wc_order_addresses
+				WHERE country=%s AND state IN {$states_as_comma_separated}
```
Same here with interpolated vs prepared.

peterfabian reviewed (changes requested):
Thanks for all the updates!

I have only 2 orders, but the actions keep scheduling. Please let me know if you want the db dump from the affected environment. This was with COT disabled (and COT tables not even existing).

![Screenshot 2022-12-02 at 14 57 49](https://user-images.githubusercontent.com/2207451/205309344-f42b9830-13c7-4efd-8911-1ecc48f60349.png)

This is the error I'm getting:

```
WordPress database error You can't specify target table 'wp_postmeta' for update in FROM clause for query UPDATE wp_postmeta
                                SET meta_value='UA77'
                                WHERE (meta_key='_billing_state' OR meta_key='_shipping_state')
                                AND meta_value='CV'
                                AND post_id IN (
                                        SELECT post_id FROM wp_postmeta WHERE
                                        (meta_key = '_billing_country' OR meta_key='_shipping_country')
                                        AND meta_value='UA'
                                )
                                LIMIT 100
```

Apparently, MySQL doesn't support this. 
> You cannot update a table and select directly from the same table in a subquery. You can work around this by using a multi-table update in which one of the tables is derived from the table that you actually wish to update, and referring to the derived table using an alias. 

https://dev.mysql.com/doc/refman/5.6/en/update.html

Konamiman committed:
Change the update statement for the post_meta table in migrate_country_states_for_orders.

Change needed because MySQL doesn't support updating a table
and selecting directly from the same table in a subquery:
https://dev.mysql.com/doc/refman/5.6/en/update.html

peterfabian commented on plugins/woocommerce/src/Database/Migrations/MigrationHelper.php:
```diff
@@ -244,7 +243,7 @@ private static function migrate_country_states_for_orders( string $country_code,
 		}
 
-		return $wpdb->get_var( $more_exist_query ) !== 0;
+		return intval( $wpdb->get_var( $more_exist_query ) ) !== 0;
```
`(int)` is preferred

https://core.trac.wordpress.org/ticket/42918

Konamiman replied:
👍🏻 Changed

peterfabian commented on plugins/woocommerce/src/Database/Migrations/MigrationHelper.php:
```diff
@@ -175,7 +175,7 @@ private static function migrate_country_states_for_orders( string $country_code,
 		 * @since 7.2.0
 		 */
-		$limit      = apply_filters( 'woocommerce_migrate_country_states_for_orders_batch_size', 100, $country_code, $old_to_new_states_mapping );
+		$limit      = apply_filters( 'woocommerce_migrate_country_states_for_orders_batch_size', 5, $country_code, $old_to_new_states_mapping );
```
Was this limit change intentional or was it only for testing?

Konamiman replied:
It wasn't intentional at all, sorry. I've fixed it.

peterfabian reviewed (changes requested):
I'm still seeing a DB error:
```
WordPress database error Incorrect usage of UPDATE and LIMIT for query UPDATE /*+ NO_MERGE(posts_with_address_in_country) */ wp_postmeta,
                                        (SELECT DISTINCT post_id FROM wp_postmeta
                                        WHERE (meta_key = '_billing_country' OR meta_key='_shipping_country') AND meta_value='NZ')
                                        AS posts_with_address_in_country
                                SET meta_value='STL'
                                WHERE (meta_key='_billing_state' OR meta_key='_shipping_state')
                                AND meta_value='SL'
                                AND wp_postmeta.post_id = posts_with_address_in_country.post_id
                                LIMIT 5 made by do_action_ref_array('action_scheduler_run_queue')
```

Konamiman committed:
Split the postmeta update in migrate_country_states_for_store_location

Split in a select + an update. This is needed because MySQL
doesn't support the LIMIT keyword in multi-table UPDATE statements.

peterfabian reviewed (approved):
Tested with CPT and HPOS on MySQL and it works fine as far as I can see. I suppose you tested this on MariaDB, so I think we should be good to go.

peterfabian commented:
Tested on a site with 10k orders, all seems to work well. Just noting that address_index fields are excluded in the update, but they are trickier to update and my guess would be that not many customers search by state code, so it should be fine.

peterfabian commented:
Also related to https://github.com/woocommerce/woocommerce/issues/32520
We need to run migration also when adding states, not only when updating them.
